### PR TITLE
Make kubectl apply create resources if not found

### LINKED ISSUE #16797: kubectl apply should create resources
#16401 

kubectl apply should create resource when it doesn't already exist, instead of returning resource not found.

### PR DESCRIPTION
**Pull request by janetkuo**
Fixes #16797

cc @kubernetes/kubectl

### CODE REVIEW

**Comment by bgrant0607** [comment]
cc @jackgr @jlowdermilk

**Comment by bgrant0607** [comment]
cc @jackgr @jlowdermilk

File: pkg/kubectl/cmd/apply.go [hunk 1]
```diff
@@ -119,7 +121,14 @@ func RunApply(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *Ap
 		}
 
 		if err := info.Get(); err != nil {
-			return cmdutil.AddSourceToErr(fmt.Sprintf("retrieving current configuration of:\n%v\nfrom server for:", info), info.Source, err)
+			if errors.IsNotFound(err) {
+				// Create the resource if it doesn't exist
+				if err := CreateAndRefresh(info); err != nil {
+					return nil
+				}
+			} else {
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/apply.go]: Suggest inverting this condition 

    if !errors.IsNotFound(err) {
        ...
    }

so you can return early and avoid the else

File: pkg/kubectl/cmd/create.go [hunk 2]
```diff
@@ -165,3 +163,13 @@ func makePortsString(ports []api.ServicePort, useNodePort bool) string {
 	}
 	return strings.Join(pieces, ",")
 }
+
+// CreateAndRefresh creates an object from input info and refreshes info with that object
+func CreateAndRefresh(info *resource.Info) error {
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/create.go]: Doesn't need to be public.

**Comment by smarterclayton** [comment]
Needs unit test.

File: pkg/kubectl/cmd/apply.go [hunk 3]
```diff
@@ -119,7 +121,14 @@ func RunApply(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *Ap
 		}
 
 		if err := info.Get(); err != nil {
-			return cmdutil.AddSourceToErr(fmt.Sprintf("retrieving current configuration of:\n%v\nfrom server for:", info), info.Source, err)
+			if errors.IsNotFound(err) {
+				// Create the resource if it doesn't exist
+				if err := CreateAndRefresh(info); err != nil {
```
Comment (deads2k) [inline pkg/kubectl/cmd/apply.go]: Doesn't creating the object based on the input obviate the need to run the three-way merge later on?

Comment (jlowdermilk) [inline pkg/kubectl/cmd/apply.go]: Returning nil skips that, but should probably call `cmdutil.PrintSuccess`, with "created" before doing so.

**Comment by janetkuo** [comment]
Thanks! Comments addressed. Add a unit test and tests in test-cmd.sh. PTAL.

**Comment by jlowdermilk** [comment]
lgtm. Squash and I'll add label.

**Comment by janetkuo** [comment]
@jlowdermilk commits squashed. Thanks!

**Comment by janetkuo** [comment]
@k8s-bot test this

**Comment by janetkuo** [comment]
Code rebased. Removing tag...

**Pull request merged**